File: Structural/Lab12/Typescript/src/Entities/house.ts
import { RealEstate } from './real_estate';
import { Customer } from './customer';
import { Realtor } from './realtor';
import { RealEstateFlyweight } from './real_state_flyweight';

export class House extends RealEstate {
    public garden: boolean;
    public pool: boolean;
    public gym: boolean;
    public heatingType: string;

    constructor();
    constructor(address: string, description: string, area: number, price: number, status: string);
    constructor(address: string, description: string, area: number, price: number, status: string, intrinsicState: RealEstateFlyweight);
    constructor(address?: string, description?: string, area?: number, price?: number, status?: string, instrinsicState?: RealEstateFlyweight) {
        if (address && description && area && price && status) {
            super(address, description, area, price, status);
            if (instrinsicState) {
                this.intrinsicState = instrinsicState;
            }
        } else {
            super("Dungeon", "tra ta ta", 50, 1000, "New");
        }
    }

    public toString(): string {
        return "House:\n" + super.toString();
    }

    public rent(customer: Customer, startDate: Date, endDate: Date): void {
        if (this.isRentable)
            console.log(`Customer ${customer.nameAndSurname} rented house on address ${this.address} from ${startDate} to ${endDate}.`);
    }

    public sell(customer: Customer): void {
        if (this.isSellable)
            console.log(`Customer ${customer.nameAndSurname} bought house on address ${this.address}.`);
    }

    public setEstateToBuy(customer: Customer, price: number): void {
        if (this.isBuyable) {
            console.log(`Customer ${customer.nameAndSurname} put up for sale house on address ${this.address} with price(USD) ${price}.`);
            this.priceInUSD = price;
            this.isBuyable = false;
        }
    }

    public clone(): House {
        const clone = Object.assign(Object.create(Object.getPrototypeOf(this)), this);
        clone.realtor = this.realtor ? Realtor.newRealtorFromObject(this.realtor) : null;
        return clone;
    }

    public shallowCopy(): House {
        return Object.assign(Object.create(Object.getPrototypeOf(this)), this);
    }

    public reset(): void {
        super.reset();
        this.heatingType = "";
        this.gym = false;
        this.garden = false;
        this.pool = false;
    }
}

File: Creational/Lab3/Python/house.py
from real_estate import *

class House(RealEstate):
    def __init__(self, address, description, area, price, status):
        super().__init__(address, description, area, price, status)
        self.isRentable = True
        self.isSellable = False    

    def __str__(self):
        return f"House:\n{super().__str__()}"

    def clone(self):
        clone = copy.deepcopy(self)
        clone.realtor = Realtor(self.realtor.id, self.realtor.name_and_surname, self.realtor.phone, self.realtor.email)
        return clone

    def shallow_copy(self):
        return copy.copy(self)
        
    def rent(self, customer, start_date, end_date):
        if self.isRentable:
            print(f"House rented to {customer.name_and_surname} from {start_date} to {end_date}")

    def sell(self, customer):
        if self.isSellable:
            print(f"House sold to {customer.name_and_surname}")